torch/onnx/_internal/fx/passes/type_promotion.py

```diff
@@ -0,0 +1,922 @@
+        return f"{self.namespace}.{self.op_name}".__hash__()
+
+    def __repr__(self):
+        return (
+            f"TypePromotionRule('{self.namespace}', '{self.op_name}', {self.promote_args_positions}, "
+            f"{self.promote_kwargs_names}, {self.promotion_kind})"
+        )
+
+    def __eq__(self, __value: object) -> bool:
+        if not isinstance(__value, TypePromotionRule):
+            return False
+        return (
+            self.namespace == __value.namespace
+            and self.op_name == __value.op_name
+            and self.promote_args_positions == __value.promote_args_positions
+            and self.promote_kwargs_names == __value.promote_kwargs_names
+            and self.promotion_kind == __value.promotion_kind
+        )
+
+    def is_valid(self) -> bool:
+        """Check if the rule is valid."""
+        # This always returns a module. If the module does not exist it will be created.
+        module = getattr(torch.ops, self.namespace)
+        py_op = getattr(module, self.op_name, None)
+        if py_op is None:
+            logger.warning(
+                "Cannot find op: %s in module: %s", self.op_name, self.namespace
+            )
+            return False
+        if not isinstance(py_op, torch._ops.OpOverloadPacket):
+            logger.warning(
+                "Op: torch.ops.%s.%s is not an OpOverloadPacket, got: %s",
+                self.namespace,
+                self.op_name,
+                type(py_op),
+            )
+            return False
+
+        return True
+
+    def preview_type_promotion(
+        self, args: tuple, kwargs: dict
+    ) -> TypePromotionSnapshot:
+        """Preview type promotion results for the given fx node.
+
+        Returns a TypePromotionSnapshot object that contains the promoted dtypes for
+        the arguments and the expected output dtype of the given fx node.
+        """
+
+        candidate_args = {
+            i: args[i]
+            for i in self.promote_args_positions
+            if i < len(args) and args[i] is not None
+        }
+        candidate_kwargs = {
+            name: kwargs[name]
+            for name in self.promote_kwargs_names
+            if name in kwargs and kwargs[name] is not None
+        }
+
+        computed_dtype, result_dtype = _prims_common.elementwise_dtypes(
+            *_pytree.tree_flatten(candidate_args)[0],
+            *_pytree.tree_flatten(candidate_kwargs)[0],
+            type_promotion_kind=self.promotion_kind,
+        )
+
+        return TypePromotionSnapshot(
+            {i: computed_dtype for i in candidate_args.keys()},
+            {name: computed_dtype for name in candidate_kwargs.keys()},
+            result_dtype,
+        )
+
+
+# NOTE: BELOW TABLE IS GENERATED FROM `TypePromotionRuleSetGenerator.generate_from_torch_refs`.
+# DO NOT EDIT MANUALLY !!!
+# For missing rules or discrepancies, please
+# 1. Run `pytest test/onnx/test_fx_type_promotion.py` to validate if the generated rule set is current.
```

Comment:
Should this be a "test" or a "utility"?

Reply:
Both are fine, I feel test is more convenient to use given the current code.
When we use proper codegen we can move it to a better location, or create a separate tool.